### PR TITLE
Try to demangle C++ symbols

### PR DESCRIPTION
**Pull request by palmtenor**
For C++ programs, this would make outputted stack traces look nicer.
Example: http://pastebin.com/LqT0nP67

### CODE REVIEW

File: src/python/bcc/__init__.py [hunk 1]
```diff
@@ -746,6 +748,19 @@ def _sym_cache(pid):
         return BPF._sym_caches[pid]
 
     @staticmethod
+    def _demangle(name):
+        if name not in BPF._demangle_caches:
+            try:
+                pipe = subprocess.Popen(
```
Comment (4ast) [inline src/python/bcc/__init__.py]: nice and useful, but can we do it on c++ side just calling a function instead of a process from python?

**Comment by palmtenor** [comment]
Do the demangling in C++ side instead.
I have little knowledge about Lua so I will leave that part untouched for now...
@4ast: I don't know what would be the effect if I change the `bcc_symbol->name` in-place, so I added a new field.

File: src/cc/bcc_syms.cc [hunk 2]
```diff
@@ -108,11 +111,22 @@ bool ProcSyms::resolve_addr(uint64_t addr, struct bcc_symbol *sym) {
...
   sym->offset = 0x0;
 
   for (Module &mod : modules_) {
-    if (addr >= mod.start_ && addr < mod.end_)
-      return mod.find_addr(addr, sym);
+    if (addr >= mod.start_ && addr < mod.end_) {
+      bool res = mod.find_addr(addr, sym);
+      if (sym->name) {
+        sym->demangle_name = abi::__cxa_demangle(
+          sym->name, nullptr, nullptr, nullptr
```
Comment (4ast) [inline src/cc/bcc_syms.cc]: the coding style looks a bit odd... why not in the same line? other than that it looks great!

File: src/cc/bcc_syms.cc [hunk 3]
```diff
@@ -52,13 +53,15 @@ bool KSyms::resolve_addr(uint64_t addr, struct bcc_symbol *sym) {
...
     sym->name = nullptr;
+    sym->demangle_name = nullptr;
     sym->module = nullptr;
     sym->offset = 0x0;
     return false;
   }
 
   auto it = std::upper_bound(syms_.begin(), syms_.end(), Symbol("", addr)) - 1;
   sym->name = (*it).name.c_str();
+  sym->demangle_name = sym->name;
```
Comment (palmtenor) [inline src/cc/bcc_syms.cc]: I assume no demangling are needed for kernel stack, right?

File: src/cc/bcc_syms.cc [hunk 4]
```diff
@@ -108,11 +111,22 @@ bool ProcSyms::resolve_addr(uint64_t addr, struct bcc_symbol *sym) {
...
-      return mod.find_addr(addr, sym);
+    if (addr >= mod.start_ && addr < mod.end_) {
+      bool res = mod.find_addr(addr, sym);
+      if (sym->name) {
+        sym->demangle_name = abi::__cxa_demangle(
+          sym->name, nullptr, nullptr, nullptr
+        );
+        if (!sym->demangle_name) {
+          sym->demangle_name = sym->name;
+        }
```
Comment (4ast) [inline src/cc/bcc_syms.cc]: extra { } are also unnecessary

File: src/cc/bcc_syms.cc (hunk 3 above)
Comment (4ast) [inline src/cc/bcc_syms.cc]: yeah. no demangling for kernel :)

File: src/cc/bcc_syms.cc (hunk 2 above)
Comment (palmtenor) [inline src/cc/bcc_syms.cc]: I have hard 80-char limit marker on my editor :P

**Comment by 4ast** [comment]
LGTM

File: src/lua/bcc/libbcc.lua [hunk 5]
```diff
@@ -101,6 +101,7 @@ void perf_reader_set_fd(struct perf_reader *reader, int fd);
 ffi.cdef[[
 struct bcc_symbol {
 	const char *name;
+    const char *demangle_name;
```
Comment (markdrayton) [inline src/lua/bcc/libbcc.lua]: indentation :)

**Comment by markdrayton** [comment]
As far as I can tell you also need to change https://github.com/iovisor/bcc/blob/d516406d1be4355f009d2bd6ca79c08bf015666d/src/lua/bcc/sym.lua#L28 to make the Lua side use the demangled name.

**Comment by markdrayton** [comment]
Also thanks for doings this :)

**Comment by palmtenor** [comment]
Thanks @markdrayton! Changed according to comment. Didn't locally verify Lua builds though, hoping the tests would cover:)

**Comment by 4ast** [comment]
can you please squash all of the commits into one?
github will preserve the discussion/comments and git history will be less noisy.
Thanks!

**Comment by palmtenor** [comment]
Done:)

**Comment by 4ast** [comment]
Thanks!

**Pull request merged**